### PR TITLE
vae

### PR DESCRIPTION
**Pull request by hnyu**
Before VAE only handles flattened input, now it can handle nested inputs.

### CODE REVIEW

File: alf/algorithms/vae.py [hunk 1]
```diff
@@ -15,98 +15,149 @@
...
+            AlgStep:
+                output: the latent vector `z`
+                state: empty tuple ()
+                info (VAEInfo): kld_loss
+        """
+        z, kld_loss = self._sampling_forward(time_step.observation)
+        return AlgStep(output=z, state=state, info=VAEInfo(loss=kld_loss))
+
+    def calc_loss(self, info: VAEInfo):
+        loss = torch.mean(info.loss)
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: It's better to use batched loss so that mask can be applied at RLAlgorithm

File: alf/algorithms/vae.py [hunk 2]
```diff
@@ -15,98 +15,149 @@
...
+        kl_div_loss = 0.5 * torch.sum(kl_div_loss, dim=-1)
         # reparameterization sampling: z = u + var ** 0.5 * eps
-        eps = tf.random.normal(
-            tf.shape(z_mean), dtype=tf.float32, mean=0., stddev=1.0)
-        z = z_mean + tf.exp(z_log_var * 0.5) * eps
+        eps = torch.randn(z_mean.shape)
+        z = z_mean + torch.exp(z_log_var * 0.5) * eps
         return z, self._beta * kl_div_loss
+
+    def train_step(self, time_step: TimeStep, state):
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: Doesn't need TimeStep as input.

File: alf/algorithms/vae_test.py [hunk 3]
```diff
@@ -12,313 +12,141 @@
...
+            pr_train = pr_train[idx]
+            for i in range(0, x_train.shape[0], self._batch_size):
+                optimizer.zero_grad()
+                batch = x_train[i:i + self._batch_size]
+                y_batch = y_train[i:i + self._batch_size]
+                pr_batch = torch.nn.functional.one_hot(
+                    pr_train[i:i + self._batch_size],
+                    int(z_mean_prior_network.input_tensor_spec.shape[0])).to(
+                        torch.float32)
+                z, kl_loss = encoder._sampling_forward([pr_batch, batch])
```
Comment (emailweixu) [inline alf/algorithms/vae_test.py]: should test `train_step` instead

File: alf/algorithms/vae.py [hunk 4]
```diff
@@ -15,98 +15,149 @@
...
+        """
+        Args:
+            time_step (TimeStep): input time step data, where its `observation`
+                is augmented with a prior input if there is a prior network.
```
Comment (emailweixu) [inline alf/algorithms/vae.py]: should make explicit how it is augmented: which one is first, which one is second

File: alf/algorithms/vae_test.py [hunk 5]
```diff
@@ -12,313 +12,141 @@
...
-        hist = model.fit([self.y_train, self.x_train],
-                         epochs=self.epochs,
-                         batch_size=self.batch_size,
-                         validation_data=([self.y_test, self.x_test], None))
-
-        last_val_loss = hist.history['val_loss'][-1]
-        # cvae seems have the lowest errors with the same settings
-        print("loss: ", last_val_loss)
-        self.assertTrue(30.0 < last_val_loss < 31.0)
+        self._input_spec = TensorSpec((1, ))
```
Comment (emailweixu) [inline alf/algorithms/vae_test.py]: call `super().setUp()`

File: alf/algorithms/vae_test.py [hunk 6]
```diff
@@ -12,313 +12,141 @@
...
 import os
-import tensorflow as tf
-import alf.algorithms.vae as vae
 import numpy as np
+import tempfile
 
-from tensorflow.keras.datasets import mnist
-import matplotlib.pyplot as plt
+import torch
+import torchvision.datasets as datasets
```
Comment (emailweixu) [inline alf/algorithms/vae_test.py]: not used

File: alf/algorithms/vae_test.py [hunk 7]
```diff
@@ -12,313 +12,141 @@
...
-                         validation_data=([self.y_test, self.x_test], None))
-
-        last_val_loss = hist.history['val_loss'][-1]
-        # cvae seems have the lowest errors with the same settings
-        print("loss: ", last_val_loss)
-        self.assertTrue(30.0 < last_val_loss < 31.0)
+        self._input_spec = TensorSpec((1, ))
+        self._epochs = 10
+        self._batch_size = 100
+        self._latent_dim = 1
```
Comment (emailweixu) [inline alf/algorithms/vae_test.py]: make it 2?

File: alf/algorithms/vae.py (hunk 1 above)
Comment (hnyu) [reply alf/algorithms/vae.py]: Why does this algorithm need mask? For any type of steps it can be trained.
